renku/infrastructure/git_merger.py

```diff
@@ -0,0 +1,414 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Merge strategies."""
+
+import os
+from json import JSONDecodeError
+from pathlib import Path
+from tempfile import mkdtemp
+from typing import List, NamedTuple, Optional, Union, cast
+
+from BTrees.OOBTree import BTree, TreeSet
+from deepdiff import DeepDiff
+from persistent import Persistent
+from persistent.list import PersistentList
+from zc.relation.catalog import Catalog
+
+from renku.command.command_builder.command import inject
+from renku.core import errors
+from renku.core.constant import RENKU_HOME
+from renku.core.interface.client_dispatcher import IClientDispatcher
+from renku.core.interface.database_dispatcher import IDatabaseDispatcher
+from renku.core.management.client import LocalClient
+from renku.core.util import communication
+from renku.domain_model.dataset import Dataset, Url
+from renku.domain_model.project import Project
+from renku.domain_model.workflow.plan import AbstractPlan
+from renku.infrastructure.database import Database, Index
+from renku.infrastructure.repository import Repository
+from renku.version import __version__
+
+RemoteEntry = NamedTuple(
+    "RemoteEntry", [("reference", str), ("database", Database), ("path", Path), ("repository", Repository)]
+)
+
+
+class GitMerger:
+    """Git metadata merger."""
+
+    client_dispatcher = inject.attr(IClientDispatcher)
+    database_dispatcher = inject.attr(IDatabaseDispatcher)
+
+    def merge(self, local: Path, remote: Path, base: Path) -> None:
+        """Merge two renku metadata entries together."""
+        client = self.client_dispatcher.current_client
+        self.remote_entries: List[RemoteEntry] = []
+
+        self._setup_worktrees(client)
+
+        result = None
+        self.local_database = self.database_dispatcher.current_database
+
+        try:
+            local_object = self.local_database.get_from_path(str(client.path / local))
+            try:
+                base_object: Optional[Persistent] = self.local_database.get_from_path(str(client.path / base))
+            except (errors.ObjectNotFoundError, JSONDecodeError):
+                base_object = None
+
+            for entry in self.remote_entries:
+                try:
+                    self.remote_database = entry.database
+                    remote_object = self.remote_database.get_from_path(str(client.path / remote))
+                    result = self.merge_objects(local_object, remote_object, base_object)
+                    break
```

Comment:
It's not clear to me why we should break. Can you please add a comment?

Reply:
This is solely to support an octo-merge, in which case there would be multiple remote branches. Once we found a remote branch that has everything to load the conflicted file, we don't need to try further branches. I can add a comment

Reply:
But on second thought, I guess if more than 1 branch has a change, we wouldn't want to break. I'll remove it